tensorflow/compiler/xla/debug_options_flags.cc

```diff
@@ -73,6 +73,7 @@ DebugOptions DefaultDebugOptionsIgnoringFlags() {
   opts.set_xla_gpu_deterministic_reductions(false);
   opts.set_xla_cpu_enable_xprof_traceme(false);
   opts.set_xla_gpu_unsafe_fallback_to_driver_on_ptxas_not_found(false);
+  opts.set_xla_multiheap_size_constraint_per_heap(-1);
```

Comment:
Is it always off by default? Should we retry with this strategy when we OOM?

Reply:
The multiheap mode is an opt-in (and always off by default now) as in the PR. Our customers somehow are trained to feel fine with adding flags into their training scripts to use XLA (^^").

We could consider the retry mechanism in the future. It is not very easy as the buffer assignment is done at compile time (in `GpuCompiler::RunBackend()`) but we only know it fails at runtime (when making a failing request to `bfc_allocator`.)

Of course, if the feature is highly desired, implementation efforts should not be an obstacle. One thing worth mentioning is that there will be a new cuMallocAsync (CUDA) API for memory allocation in 6-12 months. It will provide capability for de-fragmentation. However, integrating the cuMallocAsync into Tensorflow will require implementing a new subclass of Tensorflow::Allocator, instead of using BFCAllocator. I don't have visibility if Tensorflow will switch to use a new allocator using cuMallocAsync (e.g., for defragmentation) or sticking with using BFCAllocator (e.g., for less runtime overhead.)

We have discussed this within NVIDIA and we lean towards making this multiheap mode as an opt-in for now (as it solves our customers' immediate needs) until we have more visibility into the future of the Tensorflow allocator. Please feel free to let me know what you think and whether my explanation makes sense to you.